ledger/complete/ledger.go

```diff
@@ -86,19 +89,64 @@ func NewLedger(
 	return storage, nil
 }
 
+func NewSyncLedger(
+	wal realWAL.LedgerWAL,
+	capacity int,
+	metrics module.LedgerMetrics,
+	log zerolog.Logger,
+	pathFinderVer uint8) (*Ledger, error) {
+
+	l, err := NewLedger(wal, capacity, metrics, log, pathFinderVer)
+	if err != nil {
+		return nil, err
+	}
+
+	l.trieUpdateCh = make(chan *WALTrieUpdate, defaultTrieUpdateChanSize)
+	l.trieUpdateDoneCh = make(chan struct{})
+	return l, nil
+}
+
+// TrieUpdateChan returns a channel which is used to receive trie updates that needs to be logged in WALs.
+// This channel is closed when ledger component shutdowns down.
+func (l *Ledger) TrieUpdateChan() <-chan *WALTrieUpdate {
+	return l.trieUpdateCh
+}
+
+// TrieUpdateDoneChan returns a channel which is closed when there are no more WAL updates.
+// This is used to signal that it is safe to shutdown WAL component (closing opened WAL segment).
+func (l *Ledger) TrieUpdateDoneChan() chan<- struct{} {
+	return l.trieUpdateDoneCh
+}
+
 // Ready implements interface module.ReadyDoneAware
 // it starts the EventLoop's internal processing loop.
 func (l *Ledger) Ready() <-chan struct{} {
 	ready := make(chan struct{})
-	close(ready)
+	go func() {
+		defer close(ready)
+		// Start WAL component.
+		<-l.wal.Ready()
+	}()
 	return ready
 }
 
 // Done implements interface module.ReadyDoneAware
-// it closes all the open write-ahead log files.
 func (l *Ledger) Done() <-chan struct{} {
 	done := make(chan struct{})
-	close(done)
+	go func() {
+		defer close(done)
+		if l.trieUpdateCh != nil {
+			// Compactor is running in parallel.
+			// Ledger is responsible for closing trieUpdateCh channel,
+			// so Compactor can drain and process remaining updates.
+			close(l.trieUpdateCh)
+
+			// Wait for Compactor to finish all trie updates before closing WAL component.
+			<-l.trieUpdateDoneCh
```

Comment:
Would this wait forever? What guarantees that eventually there will be a message pushed to this channel?

Reply:
> Would this wait forever? What guarantees that eventually there will be a message pushed to this channel?

No because `Compactor` closes `trieUpdateDoneCh` [on shutdown](https://github.com/onflow/flow-go/blob/65f20620dc38ec35efce75e407a6ee241410c091/ledger/complete/compactor.go#L110-L111).  

Thanks for bringing this up :+1: , I think it would be safer to use defer to close `trieUpdateDoneCh`, so I pushed commit [b8797cc](https://github.com/onflow/flow-go/pull/2792/commits/b8797cc2e4564ffceae8cde1bb577cf2fd0863ba).